kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -19,44 +20,77 @@ use tracing::{debug, warn};
+    /// Fat pointer comparison must compare the raw data pointer as well as its metadata portion.
+    ///
+    /// Since vtable pointer comparison is not well defined and it has many nuances, we decided to
+    /// fail the user code performs such comparison.
+    ///
+    /// See https://github.com/model-checking/kani/issues/327 for more details.
     fn codegen_comparison_fat_ptr(
         &mut self,
         op: &BinOp,
         left: &Operand<'tcx>,
         right: &Operand<'tcx>,
     ) -> Expr {
         debug!(?op, ?left, ?right, "codegen_comparison_fat_ptr");
-        self.codegen_unimplemented(
-            &format!("Fat pointer comparison '{:?}'", op),
-            Type::Bool,
-            Location::None,
-            "https://github.com/model-checking/kani/issues/327",
-        )
+        let left_typ = self.operand_ty(left);
+        let right_typ = self.operand_ty(left);
+        assert_eq!(left_typ, right_typ, "Cannot compare pointers of different type");
+        assert!(self.is_ref_of_unsized(left_typ));
+
+        if self.is_vtable_fat_pointer(left_typ) {
+            // Codegen an assertion failure since vtable comparison is not stable.
+            let ret_type = Type::Bool;
+            let body = vec![
+                self.codegen_assert_false(
+                    PropertyClass::KaniCheck,
+                    format!("Reached unstable vtable comparison '{:?}'", op).as_str(),
+                    Location::None,
+                ),
+                // Assume false to block any further exploration of this path.
+                Stmt::assume(Expr::bool_false(), Location::None),
+                ret_type.nondet().as_stmt(Location::None).with_location(Location::None),
+            ];
+
+            Expr::statement_expression(body, ret_type).with_location(Location::None)
+        } else {
+            // Compare data pointer.
+            let left_ptr = self.codegen_operand(left);
+            let left_data = left_ptr.clone().member("data", &self.symbol_table);
+            let right_ptr = self.codegen_operand(right);
+            let right_data = right_ptr.clone().member("data", &self.symbol_table);
+            let data_cmp = comparison_expr(op, left_data.clone(), right_data.clone(), false);
+
+            // Compare the slice metadata (this logic could be adapted to compare vtable if needed).
+            let left_len = left_ptr.member("len", &self.symbol_table);
+            let right_len = right_ptr.member("len", &self.symbol_table);
+            let metadata_cmp = comparison_expr(op, left_len, right_len, false);
+
+            // Join the results.
+            match op {
+                // Only equal if both parts are equal.
+                BinOp::Eq => data_cmp.and(metadata_cmp),
+                // It is different if any is different.
+                BinOp::Ne => data_cmp.or(metadata_cmp),
+                // If data is different, only compare data.
+                // If data is equal, apply operator to metadata.
```

Comment:
Any references for this logic?

Reply:
Sure. Let me find the PR that implemented this on the rust compiler side.

Reply:
Yes, we need a reference here.